README.rst
pretext
=======

.. image:: https://travis-ci.org/moreati/b-prefix-all-the-doctests.svg
    :target: https://travis-ci.org/moreati/b-prefix-all-the-doctests

.. attention:: pretext 0.0.5 will be the final release of this module.
    I've agreed to transfer the package name to the `PreTeXt books`_ project.
    To continue using this module pin you version to `pretext<0.1`.

Write doctests with strings that work in Python 2.x *and* Python 3.x.

Just `import pretext` and call `pretext.activate()`. By default Python 3.x
`repr()` behaviour is used

.. code:: python

    >>> import pretext; pretext.activate()
    >>> b'Byte strings now have the same repr on Python 2.x & 3.x'
    b'Byte strings now have the same repr on Python 2.x & 3.x'
    >>> u'Unicode strings & nested strings work too'.split()
    ['Unicode', 'strings', '&', 'nested', 'strings', 'work', 'too']

.. _PreTeXt books: https://pretextbook.org/

The problem
-----------

Suppose you have the following functions and doctests

.. code:: python

    def textfunc():
        """
        >>> textfunc()
        u'A unicode string'
        """
        return u'A unicode string'

    def bytesfunc()
        """
        >>> bytesfunc()
        b'A byte string'
        """
        return b'A byte string'

Without pretext

- ``textfunc()`` will pass on Python 2.x, but fail on 3.x.
- ``bytesfunc()`` will fail on Python 2.x, but pass on 3.x.

This is because doctest compares the expected result (from the doc-string)
with the ``repr()`` of the returned value.

- ``repr(u'foo')`` returns ``u'foo'`` on Python 2.x, but ``'foo'`` on 3.x
- ``repr(b'bar')`` returns ``'bar'`` on Python 2.x, but ``b'bar'`` on 3.x

If the doctests are editted to remove the prefixes

.. code:: python

    def textfunc():
        """
        >>> textfunc()
        'A unicode string'
        """
        return u'A unicode string'

    def bytesfunc()
        """
        >>> bytesfunc()
        'A byte string'
        """
        return b'A byte string'

then the failures will be reversed

- ``textfunc()`` will now fail on Python 2.x, but pass on 3.x.
- ``bytesfunc()`` will now pass on Python 2.x, but fail on 3.x.

The hack
--------

Replace ``repr()`` and ``sys.displayhook`` with versions that always prefix
string literals, regardless of the Python version. Now the doctests can

- directly show the string values returned by functions/methods,
  without resorting to ``print()``, or ``.encode()`` etc
- successfully test the examples on all Python versions 

Proof of concept:

.. code:: python

    r"""
    >>> import sys
    >>> import pretext
    >>> myrepr = bar.PrefixRepr()
    >>> repr = myrepr.repr
    >>> def _displayhook(value):
    ...     if value is not None:
    ...         sys.stdout.write(myrepr.repr(value))
    >>> sys.displayhook = _displayhook
    >>> u''
    u''
    >>> b''
    b''
    >>> bytes()
    b''
    >>> b'\0'
    b'\x00'
    >>> b"'"
    b"'"
    """


Alternatives
------------

If you're ready to run screaming at the above, there are alternatives

- pytest_ provides ``#doctest: ALLOW_UNICODE`` and (from 2.9.0) ``#doctest: ALLOW_BYTES`` directives_

- `lxml`_ includes `lxml.html.usedoctest`_ and `lxml.usedoctest`_ modules for HTML and XML.

- Wrap byte-string returns in ``bytearray()``.
  ``repr(bytearray(b'abc')) == "bytearray(b'abc'))"`` on all versions of
  python that have ``bytearray()`` (2.6 onward) e.g.

  .. code:: python

       >>> bytearray(bytesfunc())
       bytearray(b'I return a byte (binary) string')

- Support Python 3.x exclusively
- Use ``print(bytesfunc().decode('ascii'))`` and choose your input values carefully
- Use ``#doctest: +SKIP``
- Use ``#doctest: +ELLIPSIS``

.. _pytest: http://pytest.org
.. _directives: http://pytest.org/latest/doctest.html
.. _lxml: https://pypi.python.org/pypi/lxml
.. _lxml.html.usedoctest: http://lxml.de/api/lxml.html.usedoctest-module.html
.. _lxml.usedoctest: http://lxml.de/api/lxml.usedoctest-module.html


CHANGES.rst
=========
Changelog
=========

0.0.5
-----

.. attention:: This will be the final release of this module.
    I've agreed to transfer the package name to the `PreTeXt books`_ project.
    To continue using this module pin you version to `pretext<0.1`.

* Documentation fixups

.. _PreTeXt books: https://pretextbook.org/

0.0.4
-----

* Fixed #2: `repr` of `bool` instances [thanks to John Vandenberg]
* Changed development status to Beta

0.0.3
-----

* Added ``pretext.activate()`` for instant gratification

0.0.2
-----

* Added ``prepr()`` convenience function. This mimics the builtin ``repr()``
  but always prefixes byte strings with `b` and textual strings with `u`.
* Added ``UnicodeRepr``/``urepr()`` and ``BytesRepr``/``brepr()``
  to mimic Python 2.x ``repr()`` and ``Python 3.x`` repr() respectively
* Added ``pdisplayhook()``, ``udisplayhook()``, ``bdisplayhook`` for
  conveniently overriding ``sys.displayhook``

0.0.1
-----

* Initial release


tests/test_xrepr.py
import sys
try:
    import unittest2 as unittest
except ImportError:
    import unittest

from hypothesis import given
from hypothesis import strategies as st

from pretext import prepr, brepr, urepr


@unittest.skipUnless(sys.version_info[0] == 2, "Python 2.x only")
class Python2TestCase(unittest.TestCase):
    @given(st.binary())
    def test_urepr_bytes(self, s):
        self.assertEqual(repr(s), urepr(s))

    @given(st.text())
    def test_urepr_unicode(self, s):
        self.assertEqual(repr(s), urepr(s))

    @given(st.lists(st.sampled_from([st.binary(), st.text()])))
    def test_urepr_in_list(self, l):
        self.assertEqual(repr(l), urepr(l))

    @given(st.binary())
    def test_brepr_bytes(self, s):
        self.assertEqual('b'+repr(s), brepr(s))

    @given(st.binary())
    def test_prepr_bytes(self, s):
        self.assertEqual('b'+repr(s), prepr(s))


@unittest.skipUnless(sys.version_info[0] == 3, "Python 3.x only")
class Python3TestCase(unittest.TestCase):

    @given(st.binary())
    def test_prepr_bytes(self, s):
        self.assertEqual(repr(s), brepr(s))

    @given(st.text())
    def test_brepr_unicode(self, s):
        self.assertEqual(repr(s), brepr(s))

    @given(st.lists(st.sampled_from([st.binary(), st.text()])))
    def test_brepr_in_list(self, l):
        self.assertEqual(repr(l), brepr(l))

    @given(st.text())
    def test_urepr_unicode(self, s):
        self.assertEqual('u'+repr(s), urepr(s))

    @given(st.text())
    def test_prepr_unicode(self, s):
        self.assertEqual('u'+repr(s), prepr(s))


class NoChangeTestCase(unittest.TestCase):

    @given(st.none())
    def test_none(self, s):
        self.assertEqual(repr(s), brepr(s))
        self.assertEqual(repr(s), prepr(s))
        self.assertEqual(repr(s), urepr(s))

    @given(st.booleans())
    def test_bool(self, s):
        self.assertEqual(repr(s), brepr(s))
        self.assertEqual(repr(s), prepr(s))
        self.assertEqual(repr(s), urepr(s))

    @given(st.integers())
    def test_int(self, s):
        self.assertEqual(repr(s), brepr(s))
        self.assertEqual(repr(s), prepr(s))
        self.assertEqual(repr(s), urepr(s))

    @given(st.floats())
    def test_float(self, s):
        self.assertEqual(repr(s), brepr(s))
        self.assertEqual(repr(s), prepr(s))
        self.assertEqual(repr(s), urepr(s))

    @given(st.complex_numbers())
    def test_complex(self, s):
        self.assertEqual(repr(s), brepr(s))
        self.assertEqual(repr(s), prepr(s))
        self.assertEqual(repr(s), urepr(s))


if __name__ == '__main__':
    unittest.main()



tests/test_pretext.py
r"""
>>> import sys
>>> import pretext
>>> myrepr = pretext.PrefixRepr()
>>> repr = myrepr.repr
>>> def _displayhook(value):
...     if value is not None:
...         sys.stdout.write(myrepr.repr(value))
>>> sys.displayhook = _displayhook
>>> u''
u''
>>> b''
b''
>>> bytes()
b''
>>> b'\0'
b'\x00'
>>> b"'"
b"'"
"""


def split(s):
    """
    >>> import sys
    >>> import pretext
    >>> myrepr = pretext.PrefixRepr()
    >>> repr = myrepr.repr
    >>> def _displayhook(value):
    ...     if value is not None:
    ...         sys.stdout.write(myrepr.repr(value))
    >>> sys.displayhook = _displayhook
    >>> repr = myrepr
    >>> split(u'a b c d e')
    [u'a', u'b', u'c', u'd', u'e']

    >>> split(b'a b c d e')
    [b'a', b'b', b'c', b'd', b'e']
    """
    return s.split()

def C(object):
    def join(self, seq):
        """
        >>> import sys
        >>> import pretext
        >>> myrepr = pretext.PrefixRepr()
        >>> repr = myrepr.repr
        >>> def _displayhook(value):
        ...     if value is not None:
        ...         sys.stdout.write(myrepr.repr(value))
        >>> sys.displayhook = _displayhook
        >>> c = C()
        >>> c.join([b'a', b'b', b'c', b'd', b'e']
        b'a!b!c!d!e'
        """
        return b'!'.join(seq)



tests/test_activate.py
"""
>>> import sys, pretext
>>> pretext.activate()
>>> bytes()
b''
>>> b'0123456789'
b'0123456789'
>>> u'0123456789'
'0123456789'
"""


def str_func():
    """
    This function's doctest does not override displayhook, thus checking
    that other doctests do not leak their sys.displayhook override.

    >>> str_func()
    'I return a native str'
    """
    return 'I return a native str'


def bytes_func():
    """
    >>> import sys, pretext
    >>> pretext.activate()
    >>> bytes_func()
    b'I return a byte string'
    """
    return b'I return a byte string'


def textual_func():
    """
    >>> import sys, pretext
    >>> pretext.activate(pretext.urepr, pretext.udisplayhook)
    >>> textual_func()
    u'I return a textual (unicode) string'
    """
    return u'I return a textual (unicode) string'

def bool_func(rv):
    """
    >>> import sys, pretext; pretext.activate()
    >>> bool_func(True)
    True
    >>> bool_func(False)
    False
    >>> 1 == 1
    True
    >>> 1 == 2
    False
    """
    return rv


foo.py
r"""
>>> import sys
>>> import bar
>>> myrepr = bar.PrefixRepr()
>>> repr = myrepr.repr
>>> def _displayhook(value):
...     if value is not None:
...         sys.stdout.write(myrepr.repr(value))
>>> sys.displayhook = _displayhook
>>> u''
u''
>>> b''
b''
>>> bytes()
b''
>>> b'\0'
b'\x00'
>>> b"'"
b"'"
"""


def split(s):
    """
    >>> import sys
    >>> import bar
    >>> myrepr = bar.PrefixRepr()
    >>> repr = myrepr.repr
    >>> def _displayhook(value):
    ...     if value is not None:
    ...         sys.stdout.write(myrepr.repr(value))
    >>> sys.displayhook = _displayhook
    >>> repr = myrepr
    >>> split(u'a b c d e')
    [u'a', u'b', u'c', u'd', u'e']

    >>> split(b'a b c d e')
    [b'a', b'b', b'c', b'd', b'e']
    """
    return s.split()

def C(object):
    def join(self, seq):
        """
        >>> import sys
        >>> import bar
        >>> myrepr = bar.PrefixRepr()
        >>> repr = myrepr.repr
        >>> def _displayhook(value):
        ...     if value is not None:
        ...         sys.stdout.write(myrepr.repr(value))
        >>> sys.displayhook = _displayhook
        >>> c = C()
        >>> c.join([b'a', b'b', b'c', b'd', b'e']
        b'a!b!c!d!e'
        """
        return b'!'.join(seq)



pretext.py
from __future__ import print_function

try:
    import builtins
except ImportError:
    import __builtin__ as builtins

try:
    import reprlib
except ImportError:
    import repr as reprlib

import sys

__all__ = [
    'activate',
    'Repr',
    'PrefixRepr',
    'BytesRepr',
    'UnicodeRepr',
    'prepr',
    'urepr',
    'pdisplayhook',
    'bdisplayhook',
    'udisplayhook',
]


class Repr(reprlib.Repr):
    """A Repr that doesn't truncate it's output

    >>> myrepr = Repr().repr
    >>> myrepr('0123456789')
    "'0123456789'"
    >>> len(myrepr('0123456789'))
    12
    >>> len(myrepr('0123456789'*1000))
    10002
    """
    def __init__(self):
        self.maxlevel = sys.maxsize
        self.maxdict = sys.maxsize
        self.maxlist = sys.maxsize
        self.maxset = sys.maxsize
        self.maxfrozenset = sys.maxsize
        self.maxdeque = sys.maxsize
        self.maxarray = sys.maxsize
        self.maxlong = sys.maxsize
        self.maxstring = sys.maxsize
        self.maxlong = sys.maxsize
        self.maxtuple = sys.maxsize
        self.maxother = sys.maxsize

    def repr_bool(self, obj, level):
        return '%s' % obj


class PrefixRepr(Repr):
    """A Repr that always prefixes any string, and doesn't truncate

    For byte strings this mirrors the behaviour of repr() in Python 3.x.
    For textual strings this mirrors the behaviour of repr() in Python 2.x.

    >>> prefix_repr = PrefixRepr().repr
    >>> prefix_repr(b'0123456789')
    "b'0123456789'"
    >>> prefix_repr(u'0123456789')
    "u'0123456789'"
    """

    def __init__(self):
        Repr.__init__(self)
        self._old_repr = repr

    def repr_str(self, obj, level):
        """Return the repr of a string, prefixed with b or u
        """
        if str is bytes:
            # Python 2.x
            return b'b%s' % (self._old_repr(obj),)
        else:
            # Python 3.x
            return u'u%s' % (self._old_repr(obj),)

    def repr_bytes(self, obj, level):
        return self._old_repr(obj)

    def repr_unicode(self, obj, level):
        return self._old_repr(obj)


class BytesRepr(PrefixRepr):
    """A Repr that prefixes byte strings, and doesn't truncate

    This mirrors the behaviour of repr() in Python 3.x.

    >>> py3_repr = BytesRepr().repr
    >>> py3_repr(b'0123456789')
    "b'0123456789'"
    >>> py3_repr(u'0123456789')
    "'0123456789'"
    """

    def repr_str(self, obj, level):
        if str is bytes:
            # Python 2.x
            return b'b%s' % (self._old_repr(obj),)
        else:
            # Python 3.x
            return self._old_repr(obj)

    def repr_unicode(self, obj, level):
        # Only Python 2.x has a builtin type called unicode
        r = self._old_repr(obj)
        assert r[0] == 'u'
        return r[1:]


class UnicodeRepr(PrefixRepr):
    """A Repr that prefixes textual (unicode) strings & doesn't truncate

    This mirrors the behaviour of repr() in Python 2.x.

    >>> py2_repr = UnicodeRepr().repr
    >>> py2_repr(b'0123456789')
    "'0123456789'"
    >>> py2_repr(u'0123456789')
    "u'0123456789'"
    """

    def repr_str(self, obj, level):
        if str is bytes:
            return self._old_repr(obj)
        else:
            return u'u%s' % (self._old_repr(obj),)

    def repr_bytes(self, obj, level):
        r = self._old_repr(obj)
        assert r[0] == 'b'
        return r[1:]


prepr = PrefixRepr().repr
brepr = BytesRepr().repr
urepr = UnicodeRepr().repr

def _displayhook(value, repr_fn):
    if value is None:
        return
    # Set '_' to None to avoid recursion
    builtins._ = None
    text = repr_fn(value)
    try:
        sys.stdout.write(text)
    except UnicodeEncodeError:
        bytes = text.encode(sys.stdout.encoding, 'backslashreplace')
        if hasattr(sys.stdout, 'buffer'):
            sys.stdout.buffer.write(bytes)
        else:
            text = bytes.decode(sys.stdout.encoding, 'strict')
            sys.stdout.write(text)
    sys.stdout.write("\n")
    builtins._ = value

def pdisplayhook(value):
    _displayhook(value, prepr)

def bdisplayhook(value):
    _displayhook(value, brepr)

def udisplayhook(value):
    _displayhook(value, urepr)

def activate(repr=brepr, displayhook=bdisplayhook):
    """Replace the builtin repr & sys.displayhook with versions that prefix

    The defaults causes the repr() of any string to be the same as it would
    on Python 3.x. So after `activate()` is called `repr(b'foo') == "b'foo'"`
    and `repr(u'bar') == "'bar'"`. Even on Python 2.x.

    >>> import pretext; pretext.activate()
    >>> b'Now your byte-string doctests can work on Python 2.6, 2.7, & 3.3+'
    b'Now your byte-string doctests can work on Python 2.6, 2.7, & 3.3+'
    >>> u"By default unicode strings aren't prefixed, just like on Python 3.x"
    "By default unicode strings aren't prefixed, just like on Python 3.x"

    >>> pretext.activate(pretext.urepr, pretext.udisplayhook)
    >>> [b'However you can choose Python 2.x', u'behaviour if you prefer']
    ['However you can choose Python 2.x', u'behaviour if you prefer']

    >>> pretext.activate(pretext.prepr, pretext.pdisplayhook)
    >>> (b'Or prefix all the strings', u'all the time')
    (b'Or prefix all the strings', u'all the time')
    """
    builtins.repr = repr
    sys.displayhook = displayhook

